src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1077,9 +1077,11 @@ intptr_t ObjectSynchronizer::FastHashCode(Thread* current, oop obj) {
 
     // Inflate the monitor to set the hash.
 
-    // An async deflation can race after the inflate() call and before we
-    // can update the ObjectMonitor's header with the hash value below.
-    monitor = inflate(current, obj, inflate_cause_hash_code);
+    // There's no need to inflate if the mark has already got a monitor.
+    // NOTE: an async deflation can race after we get the monitor and
+    // before we can update the ObjectMonitor's header with the hash
+    // value below.
```

Comment:
This is subtle.  The inflate logic will create a new inflated monitor, if the current one is being async deflated, but we can install the hash in that monitor (without it going away? It needs a handshake to go away so no.) and the logic below this comment will handle the async deflation race, which you have even if you call inflate.  Did I get this right?  Rereading the comment implies this is handled below, and in fact it is.